### PR TITLE
enh(java) support functions with nested template types

### LINKED ISSUE #2641: (Java) Can't match nested generic type identifier
**Describe the issue**

Can't match nested generic type identifier such as: `List<Atomic<Integer>>`. This will lead to a method that can't be matched correctly.

**Which language seems to have the issue?**

Java

**Are you using `highlight` or `highlightAuto`?**
`highlightAuto`

**Sample Code to Reproduce**

```java
static Optional<List<Token>> parseAll(String s) {
    var i = 0;
    var list = new LinkedList();
    while (i < s.length()) {
        var r = parse(s, i);
        if (r.isPresent()) {
            list.add(r.get().a());
            i = r.get().b();
        } else {
            return Optional.empty();
        }
    }
    return Optional.of(list);
}
```

https://jsfiddle.net/tx5jLs2a/8/

**Expected behavior**

The method should be recognized as a `hljs-function`, not a `hljs-keyword`.

**Additional context**

The code at the 3rd row of languages/java.js:
```javascript
  var JAVA_IDENT_RE = '[\u00C0-\u02B8a-zA-Z_$][\u00C0-\u02B8a-zA-Z_$0-9]*';
  var GENERIC_IDENT_RE = JAVA_IDENT_RE + '(<' + JAVA_IDENT_RE + '(\\s*,\\s*' + JAVA_IDENT_RE + ')*>)?';
```
It doesn't handle nested angle bracket, and I don't know how to solve recursive match by regular expression of javascript either;

### PR DESCRIPTION
**Pull request by joshgoebel**
Resolves #2641.

### Changes

Allows Java template types for functions to be recursive a single level:

IE (Token within List):

```
static Optional<List<Token>> parseAll(String s) {
}
```

- also lints Java
- also splits out literals and types from keywords

### Checklist
- [x] Added markup tests, or they don't apply here because...
- [x] Updated the changelog at `CHANGES.md`

### CODE REVIEW

**Review by allejo — APPROVED** [review]
Does this PR intentionally only account for one level of nesting?

image

Otherwise, lgtm

File: src/languages/java.js [hunk 1]
```diff
@@ -5,33 +5,116 @@ Category: common, enterprise
...
 
-import { NUMERIC } from "./lib/java.js";
+import {
+  NUMERIC as NUMBER
+} from "./lib/java.js";
 
+/**
+ * Allows recursive regex expressions to a given depth
+ *
+ * ie: recurRegex("(abc~~~)"", "~~~"", 2) becomes:
```
[the remark this answers is not in the document]
Comment (joshgoebel) [reply src/languages/java.js:15]: This was just comment, rot, updated the example to use regex.

File: src/languages/java.js [hunk 2]
```diff
@@ -5,33 +5,116 @@ Category: common, enterprise
...
+ * (abc(abc(abc)))
+ *
+ * @param {string} re
+ * @param {RegExp} substitution (should be a g mode regex)
+ * @param {number} depth
+ * @returns {string}``
+ */
+function recurRegex(re, substitution, depth) {
+  if (depth === -1) return "";
```
[the remark this answers is not in the document]
Comment (joshgoebel) [reply src/languages/java.js:24]: I don't think this is an improvement.  I'd rather be precise.

**Comment by joshgoebel** [comment]
> Does this PR intentionally only account for one level of nesting?

It accounts for the examples I was given.  If your example is common then we can bump the # slightly but there is no need for it to be higher than it should be.  Bump it to 2?

**Pull request merged**